File: src/main/java/com/sleepwalker/sleeplib/util/serialization/JsonUtil.java
package com.sleepwalker.sleeplib.util.serialization;

import com.google.gson.*;
import com.google.gson.stream.JsonWriter;
import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.nbt.JsonToNBT;
import net.minecraft.util.JSONUtils;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.IForgeRegistry;
import net.minecraftforge.registries.IForgeRegistryEntry;

import javax.annotation.Nonnull;
import java.awt.*;
import java.io.*;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.function.Supplier;

public final class JsonUtil {

    public static final Gson GSON = new GsonBuilder().disableHtmlEscaping().setPrettyPrinting().create();
    public static final Gson GSON_WITHOUT_PRETTY = new GsonBuilder().disableHtmlEscaping().create();

    public static void toJson(@Nonnull Gson gson, @Nonnull JsonObject jsonObject, @Nonnull File file) throws IOException {
        try(
            FileOutputStream fileOutputStream = new FileOutputStream(file);
            OutputStreamWriter writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8);
            JsonWriter jsonWriter = new JsonWriter(writer)
        ) {
            jsonWriter.setIndent("    ");
            gson.toJson(jsonObject, jsonWriter);
        }
    }

    public static void toJson(@Nonnull JsonObject jsonObject, @Nonnull File file) throws IOException {
        toJson(GSON, jsonObject, file);
    }

    @Nonnull
    public static JsonObject fromJsonAsObject(@Nonnull File file) throws IOException {
        try(
            FileInputStream fileInputStream = new FileInputStream(file);
            InputStreamReader reader = new InputStreamReader(fileInputStream, StandardCharsets.UTF_8)
        ) {
            return fromJsonAsObject(reader);
        }
    }

    @Nonnull
    public static JsonArray fromJsonAsArray(@Nonnull File file) throws IOException {
        try(
            FileInputStream fileInputStream = new FileInputStream(file);
            InputStreamReader reader = new InputStreamReader(fileInputStream, StandardCharsets.UTF_8)
        ) {
            return fromJsonAsArray(reader);
        }
    }

    @Nonnull
    public static JsonObject fromJsonAsObject(@Nonnull Gson gson, @Nonnull Reader reader) throws JsonSyntaxException, JsonIOException {
        JsonObject jsonObject = gson.fromJson(reader, JsonObject.class);
        return jsonObject == null ? new JsonObject() : jsonObject;
    }

    @Nonnull
    public static JsonObject fromJsonAsObject(@Nonnull Reader reader) throws JsonSyntaxException, JsonIOException {
        return fromJsonAsObject(GSON, reader);
    }

    @Nonnull
    public static JsonObject fromJsonAsObject(@Nonnull Gson gson, @Nonnull String json) throws JsonSyntaxException, JsonIOException {
        JsonObject jsonObject = gson.fromJson(json, JsonObject.class);
        return jsonObject == null ? new JsonObject() : jsonObject;
    }

    @Nonnull
    public static JsonObject fromJsonAsObject(@Nonnull String json) throws JsonSyntaxException, JsonIOException {
        return fromJsonAsObject(GSON, json);
    }

    @Nonnull
    public static JsonArray fromJsonAsArray(@Nonnull Gson gson, @Nonnull Reader reader) throws JsonSyntaxException, JsonIOException {
        JsonArray jsonArray = gson.fromJson(reader, JsonArray.class);
        return jsonArray == null ? new JsonArray() : jsonArray;
    }

    @Nonnull
    public static JsonArray fromJsonAsArray(@Nonnull Reader reader) throws JsonSyntaxException, JsonIOException {
        return fromJsonAsArray(GSON, reader);
    }

    @Nonnull
    public static JsonArray fromJsonAsArray(@Nonnull Gson gson, @Nonnull String json) throws JsonSyntaxException, JsonIOException {
        JsonArray jsonArray = gson.fromJson(json, JsonArray.class);
        return jsonArray == null ? new JsonArray() : jsonArray;
    }

    @Nonnull
    public static JsonArray fromJsonAsArray(@Nonnull String json) throws JsonSyntaxException, JsonIOException {
        return fromJsonAsArray(GSON, json);
    }

    @Nonnull
    public static JsonObject getObjectOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name){
        return getObjectOrCreate(jsonObject, name, JsonObject::new);
    }

    @Nonnull
    public static JsonObject getObjectOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull JsonObject def){
        return getObjectOrCreate(jsonObject, name, () -> def);
    }

    @Nonnull
    public static JsonObject getObjectOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull Supplier<JsonObject> def){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonObject()){
                return element.getAsJsonObject();
            }
        }

        JsonObject object = def.get();
        jsonObject.add(name, object);
        return object;
    }

    @Nonnull
    public static JsonArray getArrayOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name){
        return getArrayOrCreate(jsonObject, name, JsonArray::new);
    }

    @Nonnull
    public static JsonArray getArrayOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull Supplier<JsonArray> defaultValue){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonArray()){
                return element.getAsJsonArray();
            }
        }

        JsonArray def = defaultValue.get();
        jsonObject.add(name, def);
        return def;
    }

    @Nonnull
    public static String getStringOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull String def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isString()){
                    return primitive.getAsString();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    @Nonnull
    public static <E extends Enum<E>> E getEnumOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull E def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isString()){
                    try {
                        return Enum.valueOf(def.getDeclaringClass(), primitive.getAsString().toUpperCase(Locale.ROOT));
                    }
                    catch (IllegalArgumentException ignore){
                        return def;
                    }
                }
            }
        }

        jsonObject.addProperty(name, def.name().toLowerCase(Locale.ROOT));
        return def;
    }

    @Nonnull
    public static <T extends IForgeRegistryEntry<T>> T getRegistryEntryOrCreate(
        @Nonnull JsonObject jsonObject, @Nonnull String name,
        @Nonnull Supplier<IForgeRegistry<T>> registry
    ){
        return getRegistryEntryOrCreate(jsonObject, name, registry.get());
    }

    @Nonnull
    public static <T extends IForgeRegistryEntry<T>> T getRegistryEntryOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull IForgeRegistry<T> registry){

        ResourceLocation defKey = registry.getDefaultKey();
        if(defKey == null){
            throw new IllegalArgumentException(String.format("Registry %s does not have a default key", registry.getRegistryName()));
        }

        return getRegistryEntryOrCreate(jsonObject, name, registry, defKey);
    }

    @Nonnull
    public static <T extends IForgeRegistryEntry<T>> T getRegistryEntryOrCreate(
        @Nonnull JsonObject jsonObject, @Nonnull String name,
        @Nonnull IForgeRegistry<T> registry, @Nonnull ResourceLocation defKey
    ){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isString()){

                    ResourceLocation loc = ResourceLocation.tryCreate(primitive.getAsString());

                    if(loc != null){
                        T type = registry.getValue(loc);
                        if(type != null){
                            return type;
                        }
                    }
                }
            }
        }

        T def = registry.getValue(defKey);
        if(def == null){
            throw new IllegalArgumentException(String.format("Registry %s does not have a default value - %s", registry.getRegistryName(), defKey));
        }

        jsonObject.addProperty(name, defKey.toString());
        return def;
    }

    public static int getIntOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, int def){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsInt();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    public static long getLongOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, long def){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsLong();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    public static float getFloatOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, float def){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsFloat();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    public static double getDoubleOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, double def){

        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsDouble();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    @Nonnull
    public static BigInteger getBigIntegerOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull BigInteger def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsBigInteger();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    @Nonnull
    public static BigDecimal getDecimalOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull BigDecimal def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isNumber()){
                    return primitive.getAsBigDecimal();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    public static boolean getBoolOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, boolean def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isBoolean()){
                    return primitive.getAsBoolean();
                }
            }
        }

        jsonObject.addProperty(name, def);
        return def;
    }

    @Nonnull
    public static ResourceLocation getResourceLocationOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull ResourceLocation def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isString()){
                    ResourceLocation resourceLocation = ResourceLocation.tryCreate(primitive.getAsString());
                    if(resourceLocation != null){
                        return resourceLocation;
                    }
                }
            }
        }

        jsonObject.addProperty(name, def.toString());
        return def;
    }

    @Nonnull
    public static Color getColorOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull Color def){
        if(jsonObject.has(name)){
            JsonElement element = jsonObject.get(name);
            if(element.isJsonPrimitive()){
                JsonPrimitive primitive = element.getAsJsonPrimitive();
                if(primitive.isString()){
                    String val = primitive.getAsString();
                    if(val.startsWith("#")){
                        val = val.substring(1);
                        try {
                            long i = Long.parseLong(val, 16);
                            return new Color((i >> 16) & 0xFF, (i >> 8) & 0xFF, i & 0xFF, (i >> 24) & 0xFF);
                        }
                        catch (NumberFormatException ignore){}
                    }
                    return Color.decode(primitive.getAsString());
                }
                else if(primitive.isNumber()){
                    try {
                        long i = primitive.getAsLong();
                        return new Color((i >> 16) & 0xFF, (i >> 8) & 0xFF, i & 0xFF, (i >> 24) & 0xFF);
                    }
                    catch (NumberFormatException ignore){}
                }
            }
        }

        jsonObject.addProperty(name, "#" + Integer.toHexString(def.getAlpha() == 255 ? def.getRGB() & 0xFFFFFF : def.getRGB()));
        return def;
    }

    @Nonnull
    public static Properties getPropertiesOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull Supplier<Properties> defSupplier){
        if (jsonObject.has(name)) {
            JsonElement element = jsonObject.get(name);
            if(element.isJsonObject()){
                Properties properties = new Properties();
                for(Map.Entry<String, JsonElement> entry : element.getAsJsonObject().entrySet()){
                    if(entry.getValue().isJsonPrimitive()){
                        properties.put(entry.getKey(), entry.getValue().getAsString());
                    }
                }
                return properties;
            }
        }

        JsonObject def = new JsonObject();
        Properties defProperties = defSupplier.get();
        for(Map.Entry<Object, Object> entry : defProperties.entrySet()){
            def.addProperty(entry.getKey().toString(), entry.getValue().toString());
        }
        jsonObject.add(name, def);
        return defProperties;
    }

    @Nonnull
    public static ITextComponent getComponentOrCreate(@Nonnull JsonObject jsonObject, @Nonnull String name, @Nonnull ITextComponent def){
        if(jsonObject.has(name)){
            ITextComponent component = ITextComponent.Serializer.getComponentFromJson(jsonObject.get(name));
            if(component != null){
                return component;
            }
        }

        jsonObject.add(name, ITextComponent.Serializer.toJsonTree(def));
        return def;
    }

    public static int getAsIntOrCreate(@Nonnull JsonElement element, int def){
        if(element.isJsonPrimitive()){
            JsonPrimitive primitive = element.getAsJsonPrimitive();
            if(primitive.isNumber()){
                return primitive.getAsInt();
            }
        }
        return def;
    }

    @Nonnull
    public static String itemToJsonString(@Nonnull ItemStack stack){
        return JsonUtil.GSON_WITHOUT_PRETTY.toJson(itemToJson(stack));
    }

    @Nonnull
    public static JsonObject itemToJson(@Nonnull ItemStack stack){
        JsonObject jsonObject = new JsonObject();
        jsonObject.addProperty("id", Objects.requireNonNull(stack.getItem().getRegistryName()).toString());
        jsonObject.addProperty("count", stack.getCount());

        CompoundNBT tag = stack.getTag();
        if(tag != null){
            jsonObject.addProperty("nbt", tag.toString());
        }

        return jsonObject;
    }

    @Nonnull
    public static ItemStack parseItem(@Nonnull String json){
        try {
            return parseItem(JsonUtil.GSON_WITHOUT_PRETTY.fromJson(json, JsonObject.class));
        }
        catch (JsonSyntaxException ignore){
            return ItemStack.EMPTY;
        }
    }

    @Nonnull
    public static ItemStack parseItem(@Nonnull JsonObject json){
        if(!json.has("id")){
            return ItemStack.EMPTY;
        }

        JsonElement element = json.get("id");
        if(!(element.isJsonPrimitive() && element.getAsJsonPrimitive().isString())){
            return ItemStack.EMPTY;
        }

        ResourceLocation id = ResourceLocation.tryCreate(element.getAsString());
        if(id == null){
            return ItemStack.EMPTY;
        }

        Item item = ForgeRegistries.ITEMS.getValue(id);
        if(item == null){
            return ItemStack.EMPTY;
        }

        ItemStack stack = new ItemStack(item, JSONUtils.getInt(json, "count", 1));
        if(json.has("nbt")){
            com.mojang.brigadier.StringReader reader = new StringReader(JSONUtils.getString(json, "nbt", ""));
            JsonToNBT jsonToNBT = new JsonToNBT(reader);

            try {
                stack.setTag(jsonToNBT.readStruct());
            }
            catch (CommandSyntaxException ignore){
            }
        }

        return stack;
    }
}

File: src/main/kotlin/com/sleepwalker/sleeplib/util/serialization/JsonUtil.kt
package com.sleepwalker.sleeplib.util.serialization

import com.google.gson.*
import com.google.gson.stream.JsonWriter
import com.mojang.brigadier.StringReader
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.sleepwalker.sleeplib.util.ObjectResourceLocation
import net.minecraft.item.ItemStack
import net.minecraft.nbt.JsonToNBT
import net.minecraft.util.JSONUtils
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.ITextComponent
import net.minecraftforge.registries.ForgeRegistries
import net.minecraftforge.registries.IForgeRegistry
import net.minecraftforge.registries.IForgeRegistryEntry
import java.awt.Color
import java.io.*
import java.math.BigDecimal
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.util.*
import javax.annotation.Nonnull

val GSON: Gson = GsonBuilder().disableHtmlEscaping().setPrettyPrinting().create()
val GSON_WITHOUT_PRETTY: Gson = GsonBuilder().disableHtmlEscaping().create()

@Throws(IOException::class)
fun toJson(gson: Gson, jsonObject: JsonObject, file: File) {
    FileOutputStream(file).use { fileOutputStream ->
        OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8).use { writer ->
            JsonWriter(writer).use { jsonWriter ->
                jsonWriter.setIndent("    ")
                gson.toJson(jsonObject, jsonWriter)
            }
        }
    }
}

@Throws(IOException::class)
fun toJson(jsonObject: JsonObject, file: File) {
    toJson(GSON, jsonObject, file)
}

@Throws(IOException::class)
fun fromJsonAsObject(file: File): JsonObject {
    FileInputStream(file).use { fileInputStream ->
        InputStreamReader(fileInputStream, StandardCharsets.UTF_8).use { reader ->
            return fromJsonAsObject(reader)
        }
    }
}

@Throws(IOException::class)
fun fromJsonAsArray(file: File): JsonArray {
    FileInputStream(file).use { fileInputStream ->
        InputStreamReader(fileInputStream, StandardCharsets.UTF_8).use { reader ->
            return fromJsonAsArray(reader)
        }
    }
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsObject(gson: Gson, reader: Reader): JsonObject {
    val jsonObject = gson.fromJson(reader, JsonObject::class.java)
    return jsonObject ?: JsonObject()
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsObject(reader: Reader): JsonObject {
    return fromJsonAsObject(GSON, reader)
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsObject(gson: Gson, json: String): JsonObject {
    val jsonObject = gson.fromJson(json, JsonObject::class.java)
    return jsonObject ?: JsonObject()
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsObject(json: String): JsonObject {
    return fromJsonAsObject(GSON, json)
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsArray(gson: Gson, reader: Reader): JsonArray {
    val jsonArray = gson.fromJson(reader, JsonArray::class.java)
    return jsonArray ?: JsonArray()
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsArray(reader: Reader): JsonArray {
    return fromJsonAsArray(GSON, reader)
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsArray(gson: Gson, json: String): JsonArray {
    val jsonArray = gson.fromJson(json, JsonArray::class.java)
    return jsonArray ?: JsonArray()
}

@Throws(JsonSyntaxException::class, JsonIOException::class)
fun fromJsonAsArray(json: String): JsonArray {
    return fromJsonAsArray(GSON, json)
}

@JvmOverloads
fun JsonObject.getObjectOrCreate(name: String, def: () -> JsonObject = { JsonObject() }): JsonObject {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonObject) {
            return element.getAsJsonObject()
        }
    }
    val `object` = def()
    this.add(name, `object`)
    return `object`
}

@JvmOverloads
fun JsonObject.getArrayOrCreate(name: String, def: () -> JsonArray = { JsonArray() }): JsonArray {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonArray) {
            return element.getAsJsonArray()
        }
    }
    val array = def()
    this.add(name, array)
    return array
}

fun JsonObject.getStringOrCreate(name: String, def: String): String {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isString) {
                return primitive.asString
            }
        }
    }
    this.addProperty(name, def)
    return def
}

fun <E : Enum<E>> JsonObject.getEnumOrCreate(name: String, def: E): E {
    if(this.has(name)){
        val element = this.get(name)
        if(element.isJsonPrimitive){
            val primitive = element.getAsJsonPrimitive()
            if(primitive.isString){
                return try {
                    java.lang.Enum.valueOf(def.declaringJavaClass, primitive.asString);
                } catch (ignore: IllegalArgumentException){
                    def
                }
            }
        }
    }
    this.addProperty(name, def.name)
    return def
}

fun <T : IForgeRegistryEntry<T>> JsonObject.getRegistryEntryOrCreate(name: String, registry: IForgeRegistry<T>, defKey: ResourceLocation): T {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isString) {
                val loc = ResourceLocation.tryCreate(primitive.asString)
                if (loc != null) {
                    val type = registry.getValue(loc)
                    if (type != null) {
                        return type
                    }
                }
            }
        }
    }
    val def = registry.getValue(defKey) ?: throw IllegalArgumentException("Registry ${registry.registryName} does not have a default value - $defKey")
    this.addProperty(name, defKey.toString())
    return def
}

fun <T : IForgeRegistryEntry<T>> JsonObject.getRegistryEntryOrCreate(name: String, registry: IForgeRegistry<T>): T {
    val defKey = registry.defaultKey ?: throw IllegalArgumentException("Registry ${registry.registryName} does not have a default key")
    return this.getRegistryEntryOrCreate(name, registry, defKey)
}

fun <T : IForgeRegistryEntry<T>> JsonObject.getRegistryEntryOrCreate(name: String, registry: () -> IForgeRegistry<T>): T {
    return this.getRegistryEntryOrCreate(name, registry())
}

fun getIntOrCreate(jsonObject: JsonObject, name: String?, def: Int): Int {
    if (jsonObject.has(name)) {
        val element = jsonObject[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asInt
            }
        }
    }
    jsonObject.addProperty(name, def)
    return def
}

fun getLongOrCreate(jsonObject: JsonObject, name: String?, def: Long): Long {
    if (jsonObject.has(name)) {
        val element = jsonObject[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asLong
            }
        }
    }
    jsonObject.addProperty(name, def)
    return def
}

fun getFloatOrCreate(jsonObject: JsonObject, name: String?, def: Float): Float {
    if (jsonObject.has(name)) {
        val element = jsonObject[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asFloat
            }
        }
    }
    jsonObject.addProperty(name, def)
    return def
}

fun getDoubleOrCreate(jsonObject: JsonObject, name: String?, def: Double): Double {
    if (jsonObject.has(name)) {
        val element = jsonObject[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asDouble
            }
        }
    }
    jsonObject.addProperty(name, def)
    return def
}


fun JsonObject.getBigIntegerOrCreate(name: String, def: BigInteger): BigInteger {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asBigInteger
            }
        }
    }
    this.addProperty(name, def)
    return def
}


fun JsonObject.getDecimalOrCreate(name: String, def: BigDecimal): BigDecimal {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isNumber) {
                return primitive.asBigDecimal
            }
        }
    }
    this.addProperty(name, def)
    return def
}

fun JsonObject.getBoolOrCreate(name: String, def: Boolean): Boolean {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isBoolean) {
                return primitive.asBoolean
            }
        }
    }
    this.addProperty(name, def)
    return def
}


fun JsonObject.getResourceLocationOrCreate(name: String, def: ResourceLocation): ResourceLocation {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isString) {
                val resourceLocation = ResourceLocation.tryCreate(primitive.asString)
                if (resourceLocation != null) {
                    return resourceLocation
                }
            }
        }
    }
    this.addProperty(name, def.toString())
    return def
}

fun JsonObject.getObjectResourceLocationOrCreate(name: String, def: ObjectResourceLocation): ObjectResourceLocation {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isString) {
                val objectResourceLocation = ObjectResourceLocation.tryCreate(primitive.asString)
                if (objectResourceLocation != null) {
                    return objectResourceLocation
                }
            }
        }
    }
    this.addProperty(name, def.toString())
    return def
}

fun JsonObject.getColorOrCreate(name: String, @Nonnull def: Color): Color {
    if (this.has(name)) {
        val element = this[name]
        if (element.isJsonPrimitive) {
            val primitive = element.getAsJsonPrimitive()
            if (primitive.isString) {
                var `val` = primitive.asString
                if (`val`.startsWith("#")) {
                    `val` = `val`.substring(1)
                    try {
                        var i = `val`.toLong(16)
                        if (i <= 0xFFFFFFL) {
                            i += 0xFF000000L
                        }
                        return Color(
                            (i shr 16 and 0xFFL).toInt(),
                            (i shr 8 and 0xFFL).toInt(),
                            (i and 0xFFL).toInt(),
                            (i shr 24 and 0xFFL).toInt()
                        )
                    } catch (ignore: NumberFormatException) {
                    }
                }
                return Color.decode(primitive.asString)
            } else if (primitive.isNumber) {
                try {
                    var i = primitive.asLong
                    if (i <= 0xFFFFFFL) {
                        i += 0xFF000000L
                    }
                    return Color(
                        (i shr 16 and 0xFFL).toInt(),
                        (i shr 8 and 0xFFL).toInt(),
                        (i and 0xFFL).toInt(),
                        (i shr 24 and 0xFFL).toInt()
                    )
                } catch (ignore: NumberFormatException) {
                }
            }
        }
    }
    this.addProperty(name, "#" + Integer.toHexString(if (def.alpha == 255) def.rgb and 0xFFFFFF else def.rgb))
    return def
}

fun JsonObject.getPropertiesOrCreate(jsonObject: JsonObject, name: String, def: () -> Properties): Properties {
    if (jsonObject.has(name)) {
        val element = jsonObject[name]
        if (element.isJsonObject) {
            val properties = Properties()
            for ((key, value) in element.getAsJsonObject().entrySet()) {
                if (value.isJsonPrimitive) {
                    properties[key] = value.asString
                }
            }
            return properties
        }
    }
    val defObj = JsonObject()
    val defProperties = def()
    for ((key, value) in defProperties) {
        defObj.addProperty(key.toString(), value.toString())
    }
    jsonObject.add(name, defObj)
    return defProperties
}

fun JsonObject.getComponentOrCreate(name: String, def: ITextComponent): ITextComponent {
    if (this.has(name)) {
        val component: ITextComponent? = ITextComponent.Serializer.getComponentFromJson(this[name])
        if (component != null) {
            return component
        }
    }
    this.add(name, ITextComponent.Serializer.toJsonTree(def))
    return def
}

fun JsonElement.getAsIntOrCreate(def: Int): Int {
    if (this.isJsonPrimitive) {
        val primitive = this.getAsJsonPrimitive()
        if (primitive.isNumber) {
            return primitive.asInt
        }
    }
    return def
}

fun ItemStack.itemToJson(): JsonObject {
    val jsonObject = JsonObject()
    jsonObject.addProperty("id", Objects.requireNonNull(this.item.registryName).toString())
    jsonObject.addProperty("count", this.count)
    val tag = this.tag
    if (tag != null) {
        jsonObject.addProperty("nbt", tag.toString())
    }
    return jsonObject
}

fun ItemStack.itemToJsonString(): String {
    return GSON_WITHOUT_PRETTY.toJson(this.itemToJson())
}

fun JsonObject.getItemStackOrEmpty(): ItemStack {
    if (!this.has("id")) {
        return ItemStack.EMPTY
    }
    val element = this["id"]
    if (!(element.isJsonPrimitive && element.getAsJsonPrimitive().isString)) {
        return ItemStack.EMPTY
    }
    val id = ResourceLocation.tryCreate(element.asString) ?: return ItemStack.EMPTY
    val item = ForgeRegistries.ITEMS.getValue(id) ?: return ItemStack.EMPTY
    val stack = ItemStack(item, JSONUtils.getInt(this, "count", 1))
    if (this.has("nbt")) {
        val reader = StringReader(JSONUtils.getString(this, "nbt", ""))
        val jsonToNBT = JsonToNBT(reader)
        try {
            stack.setTag(jsonToNBT.readStruct())
        } catch (ignore: CommandSyntaxException) {
        }
    }
    return stack
}

fun parseItemStackOrEmpty(json: String): ItemStack {
    return try {
        GSON_WITHOUT_PRETTY.fromJson(json, JsonObject::class.java).getItemStackOrEmpty()
    } catch (ignore: JsonSyntaxException) {
        ItemStack.EMPTY
    }
}